feat(image): add extractRGBAs

tooooo1 opened the pull request:
## Summary

<!-- Please summarize your changes. -->

<!-- Please link to any applicable information (forum posts, bug reports, etc.). -->

a new feature in `@fepack/image` which allows users to extract a palette of colors from a given image URL. The extraction can be customized based on user-defined parameters such as the number of colors to extract and the quality of extraction.

![code](https://github.com/fepack/image/assets/77133565/0038c307-971d-417b-8f1e-df54b4957a83)

<img width="789" alt="스크린샷 2023-09-22 오전 12 56 18" src="https://github.com/fepack/image/assets/77133565/24d2fa38-1749-4b51-8f4e-0b1ca36a204b">



## Checks

<!-- For completed items, change [ ] to [x]. -->

<!-- If you leave this checklist empty, your PR will very likely be closed. -->

Please check the following:

- [x] I have written documents and tests, if needed.

tooooo1 committed:
feat: color extractor

tooooo1 committed:
feat: unit tests for color extractor

tooooo1 committed:
Create fuzzy-impalas-applaud.md

tooooo1 committed:
chore: remove path from slash('/')

manudeli commented on packages/image/src/colorExtractor.ts:
```diff
@@ -0,0 +1,76 @@
+ * @param {string} imageUrl - The URL of the image.
+ * @param {PaletteOptions} [options] - The extraction options.
+ */
+export const colorExtractor = (imageUrl: string, options?: PaletteOptions) => {
+  return loadImageFromUrl(imageUrl)
+    .then(extractImage)
+    .then((data) => {
+      if (!data) {
+        return [];
+      }
```
I think extractColor don't need to load if user used load same url before.
So I want to suggest replacing it as parameter

```suggestion
export const extractColor = (loaded: ReturnType<typeof load>, options?: PaletteOptions) => {
  return extractColors(loaded)
    .then((data) => {
      if (!data) {
        return [];
      }
```

tooooo1 replied:
no return in the `load` function.

manudeli replied:
So It's suggestion how about make load return something

tooooo1 replied:
good. This seems like it needs a little more thought. It seems like a meaningful review. thank you.

I also registered it as an issue.
https://github.com/fepack/image/issues/35

manudeli commented on packages/image/src/colorExtractor.ts:
```diff
@@ -0,0 +1,76 @@
+    .then(extractImage)
+    .then((data) => {
+      if (!data) {
+        return [];
```
```suggestion
        throw new Error('no data')
```

tooooo1 replied:
[7e01c94](https://github.com/fepack/image/pull/33/commits/7e01c94b812df91d7598a3f85a62782ee881c0bb)

tooooo1 committed:
chore: Throw descriptive error when failing to extract pixel data

tooooo1 committed:
test: remove [] case

manudeli commented on packages/image/src/colorExtractor.ts:
```diff
@@ -0,0 +1,78 @@
+ * @param {string} imageUrl - The URL of the image.
+ * @param {PaletteOptions} [options] - The extraction options.
+ */
+export const colorExtractor = (imageUrl: string, options?: PaletteOptions) => {
```
colorExtractor is not extractor, I think verb is better to name

```suggestion
export const colorExtractor = (imageUrl: string, options?: PaletteOptions) => {
```

tooooo1 replied:
hmmmm.... How about `extractColors`?

manudeli replied:
I agreed `extractColors` before in https://github.com/fepack/image/pull/33#discussion_r1333307262

tooooo1 replied:
[e199d94](https://github.com/fepack/image/pull/33/commits/e199d94d96f06ab62623f04360b1b6865e91b7d9)

manudeli commented:
Pull Request title is needed to be renamed

tooooo1 committed:
refactor: rename colorExtractor -> extractColors

tooooo1 committed:
Merge branch 'main' into feat/color-extractor

tooooo1 committed:
Update fuzzy-impalas-applaud.md

tooooo1 committed:
Update fuzzy-impalas-applaud.md

manudeli committed:
Merge branch 'main' into feat/color-extractor

tooooo1 committed:
Merge branch 'main' into feat/color-extractor

tooooo1 committed:
refactor: move root test file extractColors

tooooo1 committed:
chore: describe test name change (extractColors function -> extractColors)

manudeli commented:
extractColors.test.ts should be extractColors.spec.ts

manudeli commented on packages/image/src/extractColors.ts:
```diff
@@ -0,0 +1,78 @@
+  return context.getImageData(0, 0, canvas.width, canvas.height).data;
+};
+
+/**
+ * Loads an image from a given URL.
+ * @param {string} url - The URL of the image to load.
+ */
+const loadImageFromUrl = (url: string) => {
+  return new Promise<HTMLImageElement>((resolve, reject) => {
+    const img = new Image();
+    img.onload = () => resolve(img);
+    img.onerror = reject;
+    img.crossOrigin = "Anonymous";
+    img.src = url;
+  });
+};
```
I think it can be replaced by `load` with some modifying itself
or load may be available refer this function. I think we need to reduce repetitive code

tooooo1 replied:
Of course! The relevant matter has already been raised as an issue #35, so can I proceed directly with the next task? I've come up with a great idea regarding the load function.

tooooo1 committed:
rename: .test.ts -> .spec.ts

manudeli commented on packages/image/src/extractColors.ts:
```diff
@@ -0,0 +1,78 @@
+ * @param {string} imageUrl - The URL of the image.
+ * @param {PaletteOptions} [options] - The extraction options.
+ */
+export const extractColors = (imageUrl: string, options?: PaletteOptions) => {
+  return loadImageFromUrl(imageUrl)
+    .then(extractImage)
+    .then((data) => {
+      if (!data) {
+        throw new Error(
+          "Failed to extract pixel data from the image. Ensure the image URL is valid and accessible.",
+        );
+      }
+
+      const pixels = filterRelevantPixels(data, options?.quality);
+      const count = options?.colorCount || pixels.length;
+      return pixels.slice(0, count);
+    });
+};
```
```suggestion
export const extractColors = (imageUrl: string, options?: PaletteOptions) =>
  loadImageFromUrl(imageUrl)
    .then(extractImage)
    .then((data) => {
      if (!data) {
        throw new Error(
          "Failed to extract pixel data from the image. Ensure the image URL is valid and accessible.",
        );
      }

      const pixels = filterRelevantPixels(data, options?.quality);
      const count = options?.colorCount || pixels.length;
      return pixels.slice(0, count);
    });
```

manudeli replied:
I wanted to suggest this every time, so I made PR to force this automatically. How about this rule? https://github.com/fepack/image/pull/41

tooooo1 replied:
[d29136d](https://github.com/fepack/image/pull/33/commits/d29136d8b0917088eabebfff82705910cd79f671) 😁

manudeli committed:
chore(eslint): eslint warn arrow-body-style (#41)

## Summary
how about eslint `arrow-body-style` warn
If we want to use this rule, I think automation by eslint is better

![Sep-24-2023
22-26-25](https://github.com/fepack/image/assets/61593290/8a7e12ee-dac3-49ad-9baa-47aaa2bce3e9)

<!-- Please summarize your changes. -->

<!-- Please link to any applicable information (forum posts, bug
reports, etc.). -->

## Checks

<!-- For completed items, change [ ] to [x]. -->

<!-- If you leave this checklist empty, your PR will very likely be
closed. -->

Please check the following:

- [x] I have written documents and tests, if needed.

manudeli commented on packages/image/src/extractColors.ts:
```diff
@@ -0,0 +1,76 @@
+}
+
+/**
+ * Extracts a palette of colors from a given image URL based on provided options.
+ * @param {string} imageUrl - The URL of the image.
+ * @param {PaletteOptions} [options] - The extraction options.
+ */
+export const extractColors = (imageUrl: string, options?: PaletteOptions) =>
+  loadImageFromUrl(imageUrl)
+    .then(extractImage)
+    .then((data) => {
+      if (!data) {
+        throw new Error(
+          "Failed to extract pixel data from the image. Ensure the image URL is valid and accessible.",
+        );
+      }
+
+      const pixels = filterRelevantPixels(data, options?.quality);
+      const count = options?.colorCount || pixels.length;
+      return pixels.slice(0, count);
+    });
+
+/**
+ * Extracts pixel data from an image.
+ * @param {HTMLImageElement} img - The image element.
+ */
+const extractImage = (img: HTMLImageElement) => {
+  const canvas = document.createElement("canvas");
+  const context = canvas.getContext("2d");
+
+  if (!context) {
+    return null;
+  }
+
+  canvas.width = img.naturalWidth;
+  canvas.height = img.naturalHeight;
+  context.drawImage(img, 0, 0);
+
+  return context.getImageData(0, 0, canvas.width, canvas.height).data;
+};
```
```suggestion
/**
 * Extracts a palette of colors from a given image URL based on provided options.
 * @param {string} imageUrl - The URL of the image.
 * @param {PaletteOptions} [options] - The extraction options.
 */
export const extractColors = (imageUrl: string, options?: PaletteOptions) =>
  loadImageFromUrl(imageUrl)
    .then(extractPixelDatas)
    .then((pixelDatas) => {
      if (!pixelDatas) {
        throw new Error(
          "Failed to extractPixelDatas from the image. Ensure the image URL is valid and accessible.",
        );
      }

      const pixels = filterRelevantPixels(pixelDatas, options?.quality);
      const count = options?.colorCount || pixels.length;
      return pixels.slice(0, count);
    });

/**
 * Extracts pixel datas from an image.
 * @param {HTMLImageElement} img - The image element.
 */
const extractPixelDatas = (img: HTMLImageElement) => {
  const canvas = document.createElement("canvas");
  const context = canvas.getContext("2d");

  if (!context) {
    return null;
  }

  canvas.width = img.naturalWidth;
  canvas.height = img.naturalHeight;
  context.drawImage(img, 0, 0);

  return context.getImageData(0, 0, canvas.width, canvas.height).data;
};
```

tooooo1 replied:
How about `extractPixels`? When naming it specifically, the term `data` doesn't seem to add much meaningful context

manudeli committed:
feat(image): change extractColors to extractRGBAs (#42)

## Summary

<!-- Please summarize your changes. -->

<!-- Please link to any applicable information (forum posts, bug
reports, etc.). -->

I changed extractColors to extractRGBAs

## Checks

<!-- For completed items, change [ ] to [x]. -->

<!-- If you leave this checklist empty, your PR will very likely be
closed. -->

Please check the following:

- [x] I have written documents and tests, if needed.

manudeli commented:
I made Pull Request to give you suggestion code https://github.com/fepack/image/pull/42

manudeli commented on .changeset/fuzzy-impalas-applaud.md:
```diff
@@ -0,0 +1,5 @@
+"@fepack/image": minor
+---
+
+feat: add add extractColors
```
```suggestion
feat: add add extractRGBAs
```

manudeli committed:
Update .changeset/fuzzy-impalas-applaud.md
